Fingerprints reduce over table-rows-sample instead of realizing all

dpsutton opened the pull request:
fixes https://github.com/metabase/metabase/issues/13288

Pass the fingerprinting transducer into the query processor to reduce over the result set rather than building up a vector of results and then fingerprinting.

One possibility is to make this rff function to `table-sample-rows` required rather than optional so that no one accidentally does this again in the future.

dpsutton committed:
Fingerprints reduce over table-rows-sample instead of realizing all

tests are failing because we test fingerprinters a lot by supplying
fake results and then reducing over those results. Now we are reducing
over them in the query processor and need to get a better testing
strategy.

I think i can stub out the query-processor to take the `:rff` and just
transduce over some fake supplied results but need to verify

dpsutton committed:
Hijack query-processor to use rff on fake data

dpsutton committed:
Docstring and make private `table-rows-sample-query`

dpsutton committed:
Fix tests to use query function rather than hijack qp

also was calling empty not empty? :(

dpsutton committed:
drivers check [ci drivers]

sbelak commented on src/metabase/db/metadata_queries.clj:
```diff
@@ -93,32 +93,47 @@
+  Options: a map of
+  `:truncation-size`: [optional] size to truncate text fields if the driver supports expressions.
+  `:rff`: [optional] a reducing function to reduce over the result set in the the query-processor rather than
+  realizing the whole collection"
```
Do we actually need both options? Currently `table-rows-sample` is only used in one place. If we ever need a realised version it might be better to have an adapter function that does the transduction.

dpsutton replied:
that's true. and its kind of a footgun to start doing table samples without it as you end up with memory pressure.

That said, the testing strategy is _WAY_ nicer by just mocking out what the table-rows-sample function returns versus hijacking the query processor to just use `rff` over the return values.

sbelak commented on test/metabase/sync/analyze/fingerprint_test.clj:
```diff
@@ -120,7 +121,8 @@
   (let [fingerprinted? (atom false)]
     (with-redefs [i/fingerprint-version->types-that-should-be-re-fingerprinted fingerprint-versions
-                  metadata-queries/table-rows-sample                           (constantly [[1] [2] [3] [4] [5]])
+                  qp/process-query                                             (fn [_ {:keys [rff]}]
```
This is reasonable all things considered, although to be honest they way I've written these tests using with-redefs gross af

sbelak reviewed (approved):
Awesome! Great job, our QP is a relatively gnarly application of transducers.

camsaul commented on test/metabase/db/metadata_queries_test.clj:
```diff
@@ -64,17 +64,14 @@
 
   (testing "substring checking"
     ;; turn off validation so we can just return the query that is emitted and check for expressions
-    (s/without-fn-validation
```
might want to remove the comment now as well

camsaul commented on src/metabase/db/metadata_queries.clj:
```diff
@@ -93,32 +93,47 @@
+
+  Options: a map of
+  `:truncation-size`: [optional] size to truncate text fields if the driver supports expressions.
+  `:rff`: [optional] a reducing function to reduce over the result set in the the query-processor rather than
+  realizing the whole collection"
```
`rff` (at least elsewhere in the QP code) isn't a reducing function, it's a function that (given initial results metadata) returns a reducing function. If it was just a reducing function it would be called `rf` (conventional Clojure variable name for a *reducing function*). `rff` instead is short for *reducing function function*.

I don't know if you're actually taking an `rf` or `rff` -- the docstring seems to suggest the former while the name suggests the latter

dpsutton replied:
right thanks. it's an `rff` the query-processor will call this with the results metadata. and that's the signature of it in fingerprint.clj. I've updated the docstring now.

```clojure
(s/defn ^:private fingerprint-table!
  [table :- i/TableInstance, fields :- [i/FieldInstance]]
  (let [rff (fn [_metadata]
              (redux/post-complete
                (f/fingerprint-fields fields)
                (fn [fingerprints]
                  (reduce (fn [count-info [field fingerprint]]
                            (cond
                              (instance? Throwable fingerprint)
                              (update count-info :failed-fingerprints inc)

                              (some-> fingerprint :global :distinct-count zero?)
                              (update count-info :no-data-fingerprints inc)

                              :else
                              (do
                                (save-fingerprint! field fingerprint)
                                (update count-info :updated-fingerprints inc))))
                          (empty-stats-map (count fingerprints))
                          (map vector fields fingerprints)))))]
    (metadata-queries/table-rows-sample table fields
                                        {:truncation-size truncation-size
                                         :rff             rff})))
```

dpsutton committed:
Docstring cleanup for rff and remove outdated comment

camsaul commented on src/metabase/db/metadata_queries.clj:
```diff
@@ -93,32 +93,47 @@
-  text fields which will be honored only if the driver supports expressions."
+(defn- table-rows-sample-query
+  "Returns the mbql query to query a table for sample rows"
+  [table fields {:keys [truncation-size] :as _opts}]
```
I'm not sure why we had `:as _opts` before but it doesn't look like it's used anywhere. Not a big deal tho

dpsutton replied:
it used to just be destructured out but now the query builder is a separate function so the whole opts are passed to that. I'm a fan of `_unused` instead of just an underscore. I think this might go against the grain of the coding style here though.

camsaul reviewed (approved):
This LGTM

camsaul commented:
@dpsutton does this close any issues? If so, make sure you mention them in the PR description (e.g. "closes #<issue number>") and tag that issue with the appropriate milestone (0.38.0 in this case). Since you're merging in to `master` this will ship as part of 0.38.0. You don't need to tag the PR itself unless there's no issue associated with the PR.

When CI is green and you have reviews approving it feel free to hit the big green merge button and then close out the issue(s) that this closes if there are any.

dpsutton commented:
ah dang. this was supposed to be into 0.37.x to help out with memory pressure issues in sync. good eye @camsaul

sbelak commented:
It's the second part of the fix for #13288

dpsutton committed:
Make rff required for table-rows-sample

now you must reduce over the reducible row set rather than realizing
the results in memory and then working with them

nemanjaglumac commented:
@dpsutton it is safe to ignore these CircleCI warnings. [The last two](https://dashboard.cypress.io/projects/a394u1/runs?branches=%5B%7B%22label%22%3A%22reducible-table-sample%22%2C%22suggested%22%3Afalse%2C%22value%22%3A%22reducible-table-sample%22%7D%5D&committers=%5B%5D&flaky=%5B%5D&page=1&status=%5B%5D&tags=%5B%5D&timeRange=%7B%22startDate%22%3A%221970-01-01%22%2C%22endDate%22%3A%222038-01-19%22%7D) Cypress runs were green for you.

dpsutton commented:
> @dpsutton it is safe to ignore these CircleCI warnings. [The last two](https://dashboard.cypress.io/projects/a394u1/runs?branches=%5B%7B%22label%22%3A%22reducible-table-sample%22%2C%22suggested%22%3Afalse%2C%22value%22%3A%22reducible-table-sample%22%7D%5D&committers=%5B%5D&flaky=%5B%5D&page=1&status=%5B%5D&tags=%5B%5D&timeRange=%7B%22startDate%22%3A%221970-01-01%22%2C%22endDate%22%3A%222038-01-19%22%7D) Cypress runs were green for you.

thanks @nemanjaglumac . Its crazy to see the entire suite of cypress failed. never seen that before. I did change the target branch from master to release-x.37.x so that may have busted the cache or thrown a wrench into it.

nemanjaglumac commented:
It failed to start with this message:
```
The run you are attempting to access is already complete and will not accept new groups.

The existing run is: https://dashboard.cypress.io/projects/a394u1/runs/2704

When a run finishes all of its groups, it waits for a configurable set of time before finally completing. You must add more groups during that time period.

The --tag flag you passed was: 
The --group flag you passed was: default
The --parallel flag you passed was: true

https://on.cypress.io/already-complete
```

Changing the target branch could be the cause.

dpsutton commented:
<img width="577" alt="image" src="https://user-images.githubusercontent.com/6377293/98393787-65dd2980-201f-11eb-9f40-3789ee2c7dd3.png">

```clojure
(tt/with-temp* [Table [table]
                  Field [field (assoc {:base_type :type/Decimal, :fingerprint_version 1} :table_id (u/get-id table))]]

    (with-redefs [i/fingerprint-version->types-that-should-be-re-fingerprinted {2 #{:type/Float}}


                  qp/process-query              (fn qp-faked [q {:keys [rff]}]
                                                  (transduce identity (rff :metadata)
                                                             (mapv vector (range 10000000))
                                                             #_(reify
                                                               clojure.lang.IReduceInit
                                                               (reduce [_ rf init]
                                                                 (loop [acc init
                                                                        n 0]
                                                                   (if (= n 10000000)
                                                                     acc
                                                                     (recur (rf acc [n]) (inc n))))))))
                  fingerprint/save-fingerprint! (fn [_field fingerprint] (println fingerprint))]
      (#'fingerprint/fingerprint-table! table [field])))
```

I _think_ this should simulate the difference. The run with `(mapv vector (range 10000000))` must realize the entire collection ahead of time and the memory usages are the two higher areas on the edges. When transducing over the `reify` (which should simulate the reducible sql results) we see the much lower spot in the middle.

dpsutton committed:
empty commit for drivers [ci drivers]
